<a href="https://colab.research.google.com/github/StephanieApaza/will_customers_purchase_your_product/blob/main/Will_customers_purchase_your_products%3F.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

In [3]:
shopping_data = pd.read_csv("online_shopping_session_data.csv")
shopping_data.head()

SessionID  Administrative  Administrative_Duration  Informational  \
0          1               0                      0.0              0   
1          2               0                      0.0              0   
2          3               0                      0.0              0   
3          4               0                      0.0              0   
4          5               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay  Weekend Month  \
0         0.20       0.20         0.0         0.0    False   Feb   
1         0.00       0.10         0.0         0.0    False   Feb   
2         0.20       0.20         0.0         0.0    False   Feb   
3         0.05       0.14         0.0         0.0    False   Feb   
4         0.02       0.05         0.0         0.0     True   Feb   

         CustomerType  Purchase  
0  Returning_Customer       0.0  
1  Returning_Customer       0.0  
2  Returning_Customer       0.0  
3  Returning_Customer       0.0  
4  Returning_Customer       0.0

In [4]:
shopping_data.shape

(12055, 15)

In [5]:
shopping_data.isna().sum()

SessionID                  0
Administrative             0
Administrative_Duration    0
Informational              0
Informational_Duration     0
ProductRelated             0
ProductRelated_Duration    0
BounceRates                0
ExitRates                  0
PageValues                 0
SpecialDay                 0
Weekend                    0
Month                      0
CustomerType               1
Purchase                   1
dtype: int64

In [6]:
shopping_data.describe()

SessionID  Administrative  Administrative_Duration  Informational  \
count  12055.000000    12055.000000             12055.000000   12055.000000   
mean    6028.000000        2.320614                80.735481       0.505268   
std     3480.123082        3.320123               175.766025       1.270637   
min        1.000000        0.000000                 0.000000       0.000000   
25%     3014.500000        0.000000                 0.000000       0.000000   
50%     6028.000000        1.000000                 8.000000       0.000000   
75%     9041.500000        4.000000                93.500000       0.000000   
max    12055.000000       27.000000              3398.750000      24.000000   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
count            12055.000000    12055.000000             12055.000000   
mean                34.648971       31.688096              1191.743969   
std                141.574000       44.371672              1912.521630   
min                  0.000000        0.000000                 0.000000   
25%                  0.000000        7.000000               185.000000   
50%                  0.000000       18.000000               599.333333   
75%                  0.000000       38.000000              1464.104809   
max               2549.375000      705.000000             63973.522230   

        BounceRates     ExitRates    PageValues    SpecialDay      Purchase  
count  12055.000000  12055.000000  12055.000000  12055.000000  12054.000000  
mean       0.022125      0.042977      5.793676      0.062829      0.154223  
std        0.048277      0.048468     18.074742      0.200955      0.361177  
min        0.000000      0.000000      0.000000      0.000000      0.000000  
25%        0.000000      0.014286      0.000000      0.000000      0.000000  
50%        0.003139      0.025068      0.000000      0.000000      0.000000  
75%        0.016947      0.050000      0.000000      0.000000      0.000000  
max        0.200000      0.200000    361.763742      1.000000      1.000000

In [7]:
shopping_data.dtypes

SessionID                    int64
Administrative               int64
Administrative_Duration    float64
Informational                int64
Informational_Duration     float64
ProductRelated               int64
ProductRelated_Duration    float64
BounceRates                float64
ExitRates                  float64
PageValues                 float64
SpecialDay                 float64
Weekend                       bool
Month                       object
CustomerType                object
Purchase                   float64
dtype: object

# **Validation**

**Numeric columns**

In [8]:
# Numeric columns
numeric = shopping_data.select_dtypes(include=['number'])

for col in numeric.columns:
    try:
        # Convert to numeric
        pd.to_numeric(numeric[col], errors='raise')
    except Exception:
        # If failed, show the problematic rows
        error = pd.to_numeric(numeric[col], errors='coerce').isna() & numeric[col].notna()
        if error.any():
            print(f"Columna '{col}' tiene valores no numéricos en filas:", numeric.index[mask].tolist())

In [9]:
numeric = shopping_data.select_dtypes(include=['number'])
for col in numeric.columns:
  print(f"Number of unique values in {col} column:", numeric[col].nunique())

Number of unique values in SessionID column: 12055
Number of unique values in Administrative column: 26
Number of unique values in Administrative_Duration column: 3285
Number of unique values in Informational column: 17
Number of unique values in Informational_Duration column: 1240
Number of unique values in ProductRelated column: 306
Number of unique values in ProductRelated_Duration column: 9376
Number of unique values in BounceRates column: 1841
Number of unique values in ExitRates column: 4694
Number of unique values in PageValues column: 2641
Number of unique values in SpecialDay column: 6
Number of unique values in Purchase column: 2


In [16]:
shopping_data['Purchase'].unique()

array([ 0.,  1., nan])

**Object columns**

In [10]:
# Unique values in no numeric columns
non_numeric = shopping_data.select_dtypes('object')
for col in non_numeric.columns:
  print(f"Number of unique values in {col} column:", non_numeric[col].nunique())

Number of unique values in Month column: 11
Number of unique values in CustomerType column: 2


In [12]:
shopping_data['Month'].unique()

array(['Feb', 'Mar', 'May', 'Oct', 'June', 'Jul', 'Aug', 'Nov', 'Sep',
       'Dec', 'N'], dtype=object)

In [13]:
shopping_data['Month'].value_counts()

Month
May     3364
Nov     2861
Mar     1907
Dec     1589
Oct      549
Sep      448
Aug      433
Jul      432
June     287
Feb      184
N          1
Name: count, dtype: int64

In [14]:
shopping_data['CustomerType'].unique()

array(['Returning_Customer', 'New_Customer', nan], dtype=object)

In [15]:
shopping_data['CustomerType'].value_counts()

CustomerType
Returning_Customer    10386
New_Customer           1668
Name: count, dtype: int64

**Duplicates values**

In [17]:
# If there are duplicates
shopping_data.duplicated().sum()

np.int64(0)

In [18]:
shopping_data.duplicated(subset=['SessionID']).sum()

np.int64(0)

**Missing values**

In [19]:
missing_customertype_purchase = shopping_data[shopping_data['CustomerType'].isna()]
missing_customertype_purchase

SessionID  Administrative  Administrative_Duration  Informational  \
12054      12055               1                     14.5              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
12054                     0.0              24              1277.842857   

       BounceRates  ExitRates  PageValues  SpecialDay  Weekend Month  \
12054          0.0   0.016074         0.0         0.0     True     N   

      CustomerType  Purchase  
12054          NaN       NaN

In [20]:
# Dropping NaN values
shopping_data = shopping_data.dropna()

In [21]:
# Replacing June by Jun
shopping_data['Month'] = shopping_data['Month'].str.replace("June","Jun")

In [22]:
shopping_data.shape

(12054, 15)

Before cleaning and validation, the dataset contained 12055 rows and 15 columns. I validated all columns against the criteria in the dataset description.

- SessionID: 12055 numeric unique values consistent with the description. No cleaning was needed.
- Administrative: 26 numeric different values. No cleaning was needed.
- Administrative_duration: 3285 numeric different values. No cleaning was needed.
- Informational: 17 numeric different values. No cleaning was needed.
- Informational_Duration: 1240 numeric different values. No cleaning was needed.
- ProductRelated: 306 numeric different values. No cleaning was needed.
- ProductRelated_Duration: 9376 numeric different values. No cleaning was needed.
- BounceRates: 1841 numeric different values. No cleaning was needed.
- ExitRates: 4694 numeric different values. No cleaning was needed.
PageValues: 2641 numeric different values. No cleaning was needed.
- SpecialDay: 6 numeric different values. No cleaning was needed.
- Weekday: 2 different boolean values. No cleaning was needed.
- Month: 11 different object values, cleaning was needed to replace June to Jun and one value "N" was drop with null values in CustomerType and Purchase due to one row with missing value is less than 5% of total dataframe.
- CustomerType: 2 different object values with one null value. It was drop.
- Purchase: 2 different numeric values with one null value. It was drop.

After validation, the **dataset contains 12054 rows and 15 columns** with no missing values.

# **Analysis**

**What are the purchase rates for online shopping sessions by customer type for November and December?**

In [25]:
nov_dec_shopping = shopping_data[shopping_data['Month'].isin(['Nov','Dec'])]
nov_dec_shopping.head(10)

SessionID  Administrative  Administrative_Duration  Informational  \
5463       5464               1                39.200000              2   
5464       5465               3                89.600000              0   
5467       5468               4               204.200000              0   
5479       5480               0                 0.000000              0   
5494       5495               0                 0.000000              0   
5497       5498              13              1013.056909              2   
5507       5508               1                86.600000              0   
5509       5510               6               545.650000              3   
5510       5511               3                94.600000              0   
5515       5516               6                82.140000              0   

      Informational_Duration  ProductRelated  ProductRelated_Duration  \
5463              120.800000               7                80.500000   
5464                0.000000              57              1721.906667   
5467                0.000000              31               652.376667   
5479                0.000000              13               710.066667   
5494                0.000000              24               968.692424   
5497              102.800000              77              2697.022104   
5507                0.000000              14               349.576190   
5509              204.346667              69              1182.432333   
5510                0.000000              16              1254.000000   
5515                0.000000              12               191.100000   

      BounceRates  ExitRates  PageValues  SpecialDay  Weekend Month  \
5463     0.000000   0.010000    0.000000         0.0     True   Nov   
5464     0.000000   0.005932  204.007949         0.0     True   Nov   
5467     0.012121   0.016162    0.000000         0.0    False   Nov   
5479     0.000000   0.007692   72.522838         0.0    False   Nov   
5494     0.000000   0.000000  106.252517         0.0    False   Nov   
5497     0.000000   0.009471    1.798670         0.0    False   Nov   
5507     0.000000   0.004762   60.437378         0.0    False   Nov   
5509     0.000685   0.002998   19.715295         0.0     True   Nov   
5510     0.000000   0.019328    0.000000         0.0     True   Nov   
5515     0.000000   0.012500    0.000000         0.0    False   Nov   

            CustomerType  Purchase  
5463        New_Customer       0.0  
5464  Returning_Customer       1.0  
5467  Returning_Customer       0.0  
5479  Returning_Customer       1.0  
5494  Returning_Customer       1.0  
5497  Returning_Customer       0.0  
5507        New_Customer       1.0  
5509  Returning_Customer       1.0  
5510        New_Customer       0.0  
5515  Returning_Customer       0.0

In [26]:
num_shopping_session = nov_dec_shopping.groupby(['CustomerType'])['Purchase'].value_counts()
num_shopping_session

CustomerType        Purchase
New_Customer        0.0          529
                    1.0          199
Returning_Customer  0.0         2994
                    1.0          728
Name: count, dtype: int64

In [27]:
# Total customer type
total_new_customer = np.sum(num_shopping_session['New_Customer'])
total_returning_customer = np.sum(num_shopping_session['Returning_Customer'])

In [28]:
# Total customer type that purchased
new_customer_purchase = num_shopping_session[('New_Customer', 1)]
returning_customer_purchase = num_shopping_session[('Returning_Customer', 1)]

In [29]:
# Rates
rate_new_customer_purchase = round(new_customer_purchase/total_new_customer,3)
rate_returning_customer_purchase = round(returning_customer_purchase/total_returning_customer,3)

In [31]:
purchase_rates = {"Returning_Customer": rate_returning_customer_purchase, "New_Customer": rate_new_customer_purchase}
purchase_rates

{'Returning_Customer': np.float64(0.196), 'New_Customer': np.float64(0.273)}

The purchases rates are:
- Returning customers: 19.6%
- New customers: 27.3%

**What is the strongest correlation in total time spent among page types by returning customers in November and December?**

In [33]:
# Only Returning Customer in November and December
customers_nov_dec_shopping = nov_dec_shopping[nov_dec_shopping['CustomerType'] == 'Returning_Customer']

In [49]:
# Correlation between Administrative_Duration and Informational_Duration
corr_admin_infor = round(customers_nov_dec_shopping['Administrative_Duration']\
                         .corr(customers_nov_dec_shopping['Informational_Duration']),3)
corr_admin_infor

np.float64(0.258)

In [50]:
# Correlation between Administrative_Duration and 'ProductRelated_Duration
corr_admin_prod = round(customers_nov_dec_shopping['Administrative_Duration']\
                        .corr(customers_nov_dec_shopping['ProductRelated_Duration']),3)
corr_admin_prod

np.float64(0.417)

In [51]:
# Correlation between Informational_Duration and 'ProductRelated_Duration
corr_infor_prod = round(customers_nov_dec_shopping['Informational_Duration']\
                        .corr(customers_nov_dec_shopping['ProductRelated_Duration']),3)
corr_infor_prod

np.float64(0.367)

In [38]:
top_correlation = {"pair":('Administrative_Duration', 'ProductRelated_Duration'), "correlation":corr_admin_prod}
top_correlation

{'pair': ('Administrative_Duration', 'ProductRelated_Duration'),
 'correlation': np.float64(0.417)}

The strongest correlation in total time spent among page types is between:
*Administrative Duration and Product Related Duration.*

**What is the likelihood of achieving at least 100 sales out of 500 online shopping sessions for the returning customers?**

In [40]:
# Current rate purchase of returning customer
rate_returning_customer_purchase

np.float64(0.196)

In [43]:
# Increase in 15% in the current rate purchase
new_rate_objetive = rate_returning_customer_purchase * 0.15 + rate_returning_customer_purchase
new_rate_objetive * 100

np.float64(22.540000000000003)

In [44]:
# Likelihood of having at leat 100 sales of 500 sessions
prob_at_least_100_sales = 1 - stats.binom.cdf(100, 500, new_rate_objetive)
prob_at_least_100_sales * 100

np.float64(90.54494133589968)

The likelihood of achieving 100 sales out of 500 online shopping sessions is: *90.54%*